Reduce number of queries to read shares in a folder

starypatyk opened the pull request:
While working on https://github.com/nextcloud/android/issues/10783 I have found that the API call to retrieve shares for files in a folder (_/ocs/v2.php/apps/files_sharing/api/v1/shares?path=xyz&reshares=true&subfiles=true_) executes 6 database queries for each file.

I have modified implementation of the **ShareAPIController::getSharesInDir** to use **OC\Share20\Manager::getSharesInDir** rather than retrieving shares separately for each file in the folder.

This significantly reduces the number of database queries and improves performance. In my tests (331 files in a folder) I got the following results (most of the time is now spent in the initialization code):

_ | Execution time [ms] | # of DB queries
--- | ---: | ---:
Before | 980 | 2021
After | 280 | 40

After implementing this change, I have noticed some changes in data returned from the API.

- Permissions for room shares returned from **RoomShareProvider::getSharesInFolder** are incorrect. I have submitted [a separate PR in nextcloud/spreed](https://github.com/nextcloud/spreed/pull/8305) to fix the issue.
- ~Original code handling this API call also returned information about incoming group shares. I do not know if this was an intended behaviour or an error in the original implementation.~ **Update** - after removing the uid check for re-shares mode in https://github.com/nextcloud/server/pull/34918/commits/d0c52e56864e466f7a07cd4499d4d4831f7ba286, the new code returns the same information in this case as the original one.

starypatyk committed:
Reduce number of queries to read shares

Signed-off-by: Dariusz Olszewski <starypatyk@users.noreply.github.com>

starypatyk committed:
An attempt to fix testGetShares unit test

Signed-off-by: Dariusz Olszewski <starypatyk@users.noreply.github.com>

starypatyk commented:
I have been able to fix failing unit tests in https://github.com/nextcloud/server/pull/34918/commits/23d06eef97b9f6f6937f9c39553397a019249f27, but I am not satisfied with this approach. :disappointed: I am afraid that I do not understand the approach to the unit tests well enough to do it better. Any suggestions are welcome.

On the other hand the [failing integration tests](https://drone.nextcloud.com/nextcloud/server/25231/42/3) seem to point to a real issue with my update. I will look into it.

artonge commented on apps/files_sharing/lib/Controller/ShareAPIController.php:
```diff
@@ -762,13 +762,12 @@ private function getSharesInDir(Node $folder): array {
-
 		/** @var \OCP\Share\IShare[] $shares */
-		$shares = array_reduce($nodes, function ($carry, $node) {
-			$carry = array_merge($carry, $this->getAllShares($node, true));
```
Looks like the `getAllShares` method can be deleted then

starypatyk replied:
Agree - corrected in https://github.com/nextcloud/server/pull/34918/commits/c51e756ab2c46b4f0ba23a73eb9c00cb7558f2f4

starypatyk commented:
I have managed to repeat locally the [first failing scenario](https://github.com/nextcloud/server/blob/master/build/integration/sharing_features/sharing-v1-part2.feature#L306) of the integration tests.

At the moment I do not know how to fix the code. :disappointed:

The problem is that:

- For [the second share expected in the result](https://github.com/nextcloud/server/blob/master/build/integration/sharing_features/sharing-v1-part2.feature#L333) the **user1** is neither the owner of the share nor the owner of the file.
- Previous code worked as expected, since in [DefaultShareProvider::getSharesBy](https://github.com/nextcloud/server/blob/master/lib/private/Share20/DefaultShareProvider.php#L774) the code retrieved all the shares for the specified node (in this case a subfolder of a folder shared originally with user1).
- There is no similar mechanism in [DefaultShareProvider::getSharesInFolder](https://github.com/nextcloud/server/blob/master/lib/private/Share20/DefaultShareProvider.php#L685).

A couple of questions:

1. Should this be handled in `ShareAPIController::getSharesInDir` or rather in `DefaultShareProvider::getSharesInFolder` (and probably in other share providers as well)?
2. Any suggestions how to efficiently retrieve required information from the database? Would it make sense to construct the database query like below? Starting from the `oc_filecache` table, column list simplified for clarity.

```
SELECT
  f.parent, f.fileid, f.path,
  s.id, s.uid_owner, s.uid_initiator, s.file_target
FROM
  oc_filecache f
  INNER JOIN oc_share s ON f.fileid = s.item_source
WHERE f.parent = folder-node-id-here;
```

starypatyk committed:
Experimental change - disable uid check

Signed-off-by: Dariusz Olszewski <starypatyk@users.noreply.github.com>

starypatyk commented:
It seems that the quick-and-dirty change in https://github.com/nextcloud/server/pull/34918/commits/d0c52e56864e466f7a07cd4499d4d4831f7ba286 fixed the failing integration tests. :open_mouth:

Do you think this approach is correct?

If yes, I will clean it up and make updates for the other review comments.

starypatyk committed:
Fixes after review

Signed-off-by: Dariusz Olszewski <starypatyk@users.noreply.github.com>

starypatyk committed:
Seems to run as expected with uid check disabled
Code clean-up

Signed-off-by: Dariusz Olszewski <starypatyk@users.noreply.github.com>

starypatyk commented on apps/files_sharing/lib/Controller/ShareAPIController.php:
```diff
@@ -762,13 +762,12 @@ private function getSharesInDir(Node $folder): array {
-		}, []);
+		$shares = [];
+		$sharesInFolder = $this->shareManager->getSharesInFolder($this->currentUser, $folder, true);
+		foreach ($sharesInFolder as $key => &$value) {
```
Agree - corrected in https://github.com/nextcloud/server/pull/34918/commits/c51e756ab2c46b4f0ba23a73eb9c00cb7558f2f4

starypatyk commented:
I have committed small changes already requested by reviewers in https://github.com/nextcloud/server/pull/34918/commits/c51e756ab2c46b4f0ba23a73eb9c00cb7558f2f4

After removing the uid check for reshares mode in **DefaultShareProvider::getSharesInFolder**, the failing tests run OK. In the last run there is a failure in **acceptance-app-files-sharing**, but this one was OK just before clean-up, so I assume that this is a random error that I have seen already in the previous PR (#34471).

I do not plan any more changes unless there are additional suggestions from review.

come-nc commented on apps/files_sharing/tests/Controller/ShareAPIControllerTest.php:
```diff
@@ -1440,6 +1440,22 @@ function ($user, $shareType, $node) use ($shares) {
+		$sharesWithoutTypes = [];
+		foreach ($shares as $file => &$fileShares) {
+			$sharesWithoutTypes[$file] = [];
+			foreach ($fileShares as $shareType => &$shareTypeShares) {
```
```suggestion
			foreach ($fileShares as $shareType => $shareTypeShares) {
```

starypatyk replied:
I have overlooked these - corrected in https://github.com/nextcloud/server/pull/34918/commits/53dc17915ea7aa6afdcfb5fd047223b8c6f3c8f1

come-nc commented:
> I have committed small changes already requested by reviewers in [c51e756](https://github.com/nextcloud/server/commit/c51e756ab2c46b4f0ba23a73eb9c00cb7558f2f4)
> 
> After removing the uid check for reshares mode in **DefaultShareProvider::getSharesInFolder**, the failing tests run OK. In the last run there is a failure in **acceptance-app-files-sharing**, but this one was OK just before clean-up, so I assume that this is a random error that I have seen already in the previous PR (#34471).
> 
> I do not plan any more changes unless there are additional suggestions from review.

How is the removal of uid check affecting the performance benchmark shown in your first post?
And how is this affecting the other places calling `getSharesInFolder` ?

starypatyk committed:
Additional fixes after review (removed references)

Signed-off-by: Dariusz Olszewski <starypatyk@users.noreply.github.com>

starypatyk commented:
@come-nc 

> How is the removal of uid check affecting the performance benchmark shown in your first post?

Same results as previously - within error margin. The timings vary between 260 and 290 ms. Even the query count is not always the same, I see a slightly varying number of `UPDATE oc_authtoken` queries during the initialization sequence.

Average results from a few runs are:

_ | Execution time [ms] | # of DB queries
-- | -- | --
Before | 980 | 2021
After | 275 | 40

> And how is this affecting the other places calling `getSharesInFolder` ?

This is a very good question. I have removed the uid check to fix the following integration tests:

```
--- Failed scenarios:

    /drone/src/build/integration/sharing_features/sharing-v1-part2.feature:306
    /drone/src/build/integration/sharing_features/sharing-v1-part2.feature:381
    /drone/src/build/integration/sharing_features/sharing-v1-part2.feature:406
```

This also fixed the issue that I have noticed at the beginning: original code handling this API call also returned information about incoming group shares. After removal of the uid check the new code behaves the same.

I suspect that there are some other scenarios when `getSharesInFolder` is going to return more data (shares in which specified user is neither owner, nor initiator). Unfortunately I am not able to assess when/how this is going to show.

szaimen commented:
Seems like the tests pass now...

starypatyk commented:
Any chances to move this PR forward? It has been submitted some time ago already... Is there anything I should/could do?

To add a bit more context - the API call I am trying to optimize (retrieve shares for files in a folder) is executed frequently by the Android app - every time the user changes a folder or returns from the file view to the file list view.

I have rebased, so merging should be straightforward. Most of the tests pass, I do not see how the failing ones might be related to my changes. Or am I mistaken?

> /drone/src/build/integration/features/contacts-menu.feature:91
> /drone/src/build/integration/features/contacts-menu.feature:146

starypatyk committed:
Revert: Experimental change - disable uid check

Signed-off-by: Dariusz Olszewski <starypatyk@users.noreply.github.com>

starypatyk commented on lib/private/Share20/DefaultShareProvider.php:
```diff
@@ -683,13 +683,6 @@ public function getSharesInFolder($userId, Folder $node, $reshares, $shallow = t
 		 */
 		if ($reshares === false) {
 			$qb->andWhere($qb->expr()->eq('uid_initiator', $qb->createNamedParameter($userId)));
-		} else {
-			$qb->andWhere(
-				$qb->expr()->orX(
-					$qb->expr()->eq('uid_owner', $qb->createNamedParameter($userId)),
-					$qb->expr()->eq('uid_initiator', $qb->createNamedParameter($userId))
-				)
-			);
```
I am also concerned about this change.

I have reverted it to check once again and I see that [three integration tests fail consistently](https://drone.nextcloud.com/nextcloud/server/33271/43/3) - in exactly the same way as when I started the PR.

````
--- Failed scenarios:

    /drone/src/build/integration/sharing_features/sharing-v1-part2.feature:306
    /drone/src/build/integration/sharing_features/sharing-v1-part2.feature:381
    /drone/src/build/integration/sharing_features/sharing-v1-part2.feature:406
````

Please see my comments above - esp. [this one](https://github.com/nextcloud/server/pull/34918#issuecomment-1302623761) and [this one](https://github.com/nextcloud/server/pull/34918#issuecomment-1306230927).

I would welcome any suggestion how to fix this in a better way.

starypatyk replied:
Hi @icewind1991,

I have looked once again into this issue. I think I have a better understanding now.

One of the failing tests describes the following scenario:
https://github.com/nextcloud/server/blob/96f01183123c7745260e0fa2fbe02cbec3a23166/build/integration/sharing_features/sharing-v1-part2.feature#L306-L318

This scenario expects two shares to be returned for **user1**: **user1 -> user2** and **user2 -> user3**. Obviously in the second share **user1** is neither owner nor initiator - but the share should be returned anyway.

I have recreated data for this scenario in my test environment and tried to repeat the issue from the UI. I have not been able to repeat this in the UI, but I have noticed a discrepancy between the following two API calls:

1. Retrieve shares for the **/PARENT (2)/CHILD** folder (used in the UI)
```
/ocs/v2.php/apps/files_sharing/api/v1/shares?path=%2FPARENT%20\(2\)%2FCHILD&reshares=true
```

2. Retrieve shares for the **/PARENT (2)** folder and its subfolders/files (used in the test)
```
/ocs/v2.php/apps/files_sharing/api/v1/shares?path=%2FPARENT%20\(2\)&reshares=true&subfiles=true
```

In the first case the shares are retrieved by **DefaultShareProvider::getSharesBy‎** that contains the following fragment to handle the **reshares** mode:
https://github.com/nextcloud/server/blob/96f01183123c7745260e0fa2fbe02cbec3a23166/lib/private/Share20/DefaultShareProvider.php#L760-L771

Here the **userId** parameter is ignored in **reshares** mode when the **node** parameter is specified.

Before my changes the second API call was handled by invoking **getSharesBy** individually for each item in the folder.

In this PR I want to reuse better handling already available in **DefaultShareProvider::getSharesInFolder**.

So, to achieve consistency with previous behaviour - I think that the **userId** check **should be removed** from the **getSharesInFolder** method. The folder node must be provided, the code always adds an additional filter to retrieve only shares relevant to files/subfolders inside the provided folder.
https://github.com/nextcloud/server/blob/96f01183123c7745260e0fa2fbe02cbec3a23166/lib/private/Share20/DefaultShareProvider.php#L696-L718

Is my understanding correct? Or have I overlooked something?

starypatyk replied:
Hi @icewind1991,

Trying once again to move this PR forward. I rebased the changes yesterday and once again removed the `uid` check in `reshares` mode. Please see my explanation in the comment above.

This is the last major item still open on the [list of performance issues in the Android app](https://github.com/nextcloud/android/issues/10783) that I started over a year ago.

Would be great to get this merged.

starypatyk committed:
Revert "Revert: Experimental change - disable uid check"

This reverts commit 1367fc1935ee23bd25ea6bf0687ba141b4d96426.

Signed-off-by: Dariusz Olszewski <starypatyk@users.noreply.github.com>

starypatyk commented:
Hi @nickvergessen and @icewind1991,

I see a new PR #42638 touching the same area as this one - i.e. the `DefaultShareProvider::getSharesInFolder` method.

Could we use this as an opportunity to get back to this PR? This is pending for a long time and is quite important to [improve performance of the Android app](https://github.com/nextcloud/android/issues/10783).

I will be glad to provide additional explanation and make further changes when suggested. At the moment I think that I had already responded to all questions/concerns and I have no idea what else I could do, to move this one forward.

nickvergessen commented:
I will bring it to the performance group and check for feedback the first week of february

starypatyk commented:
@AndyScherzinger @icewind1991 - Is there anything I should/could do regarding this PR?

I see that one of the failing tests is 'block unconventional commits'. I can squash all the commits into one if this is going to help.
